UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,736 @@
+  DEBUG ((DEBUG_INFO, "MinimalNeededSize :%x\n", MinimalNeededSize));
+
+  for (Node = FdtNextNode (Fdt, 0, &Depth); Node >= 0; Node = FdtNextNode (Fdt, Node, &Depth)) {
+    NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + Node + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
+    DEBUG ((DEBUG_INFO, "\n   Node(%08x)  %a   Depth %x", Node, NodePtr->Name, Depth));
+    // memory node
+    if (AsciiStrnCmp (NodePtr->Name, "memory@", AsciiStrLen ("memory@")) == 0) {
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+        if (AsciiStrCmp (TempStr, "reg") == 0) {
+          Data64        = (UINT64 *)(PropertyPtr->Data);
+          StartAddress  = Fdt64ToCpu (*Data64);
+          NumberOfBytes = Fdt64ToCpu (*(Data64 + 1));
+          DEBUG ((DEBUG_INFO, "\n         Property(%08X)  %a", Property, TempStr));
+          DEBUG ((DEBUG_INFO, "  %016lX  %016lX", StartAddress, NumberOfBytes));
+          if (!IsHobConstructed) {
+            if ((NumberOfBytes > MinimalNeededSize) && (StartAddress < BASE_4GB)) {
+              MemoryBottom     = StartAddress + NumberOfBytes - MinimalNeededSize;
+              FreeMemoryBottom = MemoryBottom;
+              FreeMemoryTop    = StartAddress + NumberOfBytes;
+              MemoryTop        = FreeMemoryTop;
+
+              DEBUG ((DEBUG_INFO, "MemoryBottom :0x%llx\n", MemoryBottom));
+              DEBUG ((DEBUG_INFO, "FreeMemoryBottom :0x%llx\n", FreeMemoryBottom));
+              DEBUG ((DEBUG_INFO, "FreeMemoryTop :0x%llx\n", FreeMemoryTop));
+              DEBUG ((DEBUG_INFO, "MemoryTop :0x%llx\n", MemoryTop));
+              mHobList         =  HobConstructor ((VOID *)(UINTN)MemoryBottom, (VOID *)(UINTN)MemoryTop, (VOID *)(UINTN)FreeMemoryBottom, (VOID *)(UINTN)FreeMemoryTop);
+              IsHobConstructed = TRUE;
+              NewHobList       = (UINTN)mHobList;
+              break;
+            }
+          }
+        }
+      }
+    } // end of memory node
+    else if (AsciiStrnCmp (NodePtr->Name, "pci-rb", AsciiStrLen ("pci-rb")) == 0) {
+      PropertyPtr = FdtGetProperty (Fdt, Node, "compatible", &TempLen);
+      TempStr = (CHAR8 *)(PropertyPtr->Data);
+      if (AsciiStrnCmp (TempStr, "pci-rb", AsciiStrLen ("pci-rb")) == 0) {
+        RootBridgeCount++;
+      }
+    }
+  }
+
+  index = RootBridgeCount - 1;
+  Depth = 0;
+  for (Node = FdtNextNode (Fdt, 0, &Depth); Node >= 0; Node = FdtNextNode (Fdt, Node, &Depth)) {
+    NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + Node + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
+    DEBUG ((DEBUG_INFO, "\n   Node(%08x)  %a   Depth %x", Node, NodePtr->Name, Depth));
+
+    // memory node
+    if ((AsciiStrnCmp (NodePtr->Name, "memory@", AsciiStrLen ("memory@")) == 0) && (Depth == 1)) {

```

Comment:
This specifies a `Depth` in the check too. Is it a different memory node to the ones above?

Reply:
I think this is for child nodes?

Reply:
Currently we need Depth == 1 so it will not confuse with reserved-memory->memory nodes. We can think about better way later.

Reply:
Ack.

Reply:
will re-visit this.

Reply:
resolved.